src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs

```diff
@@ -134,17 +134,46 @@ public static void AddWebJobsScriptHost(this IServiceCollection services, IConfi
             // Secret management
             services.TryAddSingleton<ISecretManagerProvider, DefaultSecretManagerProvider>();
 
-            // Shared memory data transfer and function data cache
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            // Shared memory data transfer
+            services.AddSingleton<ISharedMemoryManager>(p =>
             {
-                services.AddSingleton<IMemoryMappedFileAccessor, MemoryMappedFileAccessorWindows>();
-            }
-            else
+                var environment = p.GetService<IEnvironment>();
+                if (!environment.SupportsSharedMemoryTransfer())
+                {
+                    return new NullSharedMemoryManager();
+                }
+                else
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    {
+                        services.AddSingleton<IMemoryMappedFileAccessor, MemoryMappedFileAccessorWindows>();
+                    }
+                    else
+                    {
+                        services.AddSingleton<IMemoryMappedFileAccessor, MemoryMappedFileAccessorUnix>();
+                    }
+
+                    var fileaccessor = p.GetService<IMemoryMappedFileAccessor>();
```

Comment:
@michaelpeng36 I think I finally figured out how to debug the host on my mac and `fileaccessor` is being returned as null here. Admittedly I don't know how `AddSingleton` works, but I wonder if there's a timing/ordering issue? It seems to work if I move the `AddSingleton<IMemoryMappedFileAccessor` stuff ahead of `services.AddSingleton<ISharedMemoryManager`

Reply:
That's because IMemoryMappedFileAccessor is required when constructing ISharedMemoryManager. Order of these initializations matters.